### PR TITLE
Add elasticsearch distribution plugin

### PR DESCRIPTION
**Pull request by rjernst**
Several types of distributions are built and tested in elasticsearch,
ranging from the current version, to building or downloading snapshot or
released versions. Currently tests relying on these have to contain
logic deciding where and how to pull down these distributions.

This commit adds an distributiond download plugin for each project to
manage which versions and variants the project needs. It abstracts away
all need for knowing where a particular version comes from, like a local
distribution or bwc project, or pulling from the elastic download
service. This will be used in a followup PR by the testclusters and
vagrant tests.

### CODE REVIEW

**Comment by rjernst** [comment]
Note that this does not yet handle pulling distributions for external users of build-tools. That will require some more cleanup, for example to BwcVersions.

File: buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+            return;
+        }
+
+        if (getType().equals("archive")){
+            // defaults for archive, set here instead of via convention so integ-test-zip can verify they are not set
+            if (platform.isPresent() == false) {
+                platform.set(CURRENT_PLATFORM);
+            }
+        } else { // rpm or deb
+            if (platform.isPresent()) {
```
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java]: Wouldn't specifying "linux" technically be ok even if it's superfluous?

File: buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+        // pkg private so plugin can configure
+        final Configuration configuration;
+
+        private Extracted(Configuration configuration) {
+            this.configuration = configuration;
+        }
+
+        @Override
+        public Iterator<File> iterator() {
```
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java]: I assume this is primarily so we can do things like `from distribution.extracted`? I'm wondering if there would be any benefit having `Extracted` just implement `FileCollection` and just delegate all those to the configuration as well. Probably not. For what it's worth in Kotlin this would simply be: `class Extracted(val configuration: Configuration) : FileCollection by configuration` 😉

File: buildSrc/src/test/java/org/elasticsearch/gradle/DistributionDownloadPluginTests.java [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+        ElasticsearchDistribution distro = checkDistro(createProject(null), "testdistro", null, "archive", "linux", "oss", true);
+        assertEquals(distro.getVersion(), Version.fromString(VersionProperties.getElasticsearch()));
+    }
+
+    public void testBadVersionFormat() {
+        assertDistroError(createProject(null), "testdistro", "badversion", "archive", "linux", "oss", true,
+            "Invalid version format: 'badversion'");
+    }
+
+    public void tesTypeDefault() {
```
Comment (mark-vieira) [inline buildSrc/src/test/java/org/elasticsearch/gradle/DistributionDownloadPluginTests.java]: I believe this is a typo.

File: buildSrc/src/test/java/org/elasticsearch/gradle/DistributionDownloadPluginIT.java [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+public class DistributionDownloadPluginIT extends GradleIntegrationTestCase {
+    // TODO: check reuse of root task across projects MOVE TO UNIT TEST
+    // TODO: make extracted assert also do file assert, to reduce separate test builds
+    // TODO: future: check integ-test-zip to maven, snapshots to snapshot service for external project
+
+    private static final Pattern DISTRO_FILE_LOGLINE = Pattern.compile("DISTRO FILE: (.*)");
+    private static final Pattern DISTRO_EXTRACTED_LOGLINE = Pattern.compile("DISTRO EXTRACTED: (.*)");
+
+    public void testCurrent() throws Exception {
```
Comment (mark-vieira) [inline buildSrc/src/test/java/org/elasticsearch/gradle/DistributionDownloadPluginIT.java]: I think we should split these tests up and have a separate one for the file distro and extracted one. These tests currently run two separate builds, either of which could fail for a different reason. I think it makes more sense to have more discrete tests.

File: buildSrc/src/test/java/org/elasticsearch/gradle/DistributionDownloadPluginIT.java [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+            "tests.local_distro.project", ":distribution:bwc:minor",
+            "tests.current_version", "2.0.0");
+        assertExtractedDistro("1.1.0", "archive", "linux", null, null,
+            "tests.local_distro.config", "linux-tar",
+            "tests.local_distro.project", ":distribution:bwc:minor",
+            "tests.current_version", "2.0.0");
+    }
+
+    public void testReleased() throws Exception {
+        WireMockServer wireMock = new WireMockServer(0);
```
Comment (mark-vieira) [inline buildSrc/src/test/java/org/elasticsearch/gradle/DistributionDownloadPluginIT.java]: I love that we are doing full end-to-end integration testing here. 👍

File: buildSrc/src/test/java/org/elasticsearch/gradle/DistributionDownloadPluginIT.java [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+        }
+    }
+
+    private void assertFileDistro(String version, String type, String platform, String flavor, Boolean bundledJdk,
+                                  String... sysProps) throws IOException {
+        List<String> finalSysProps = new ArrayList<>();
+        addDistroSysProps(finalSysProps, version, type, platform, flavor, bundledJdk);
+        finalSysProps.addAll(Arrays.asList(sysProps));
+        runBuild(":subproj:getDistro", result -> {
+                Matcher matcher = DISTRO_FILE_LOGLINE.matcher(result.getOutput());
```
Comment (mark-vieira) [inline buildSrc/src/test/java/org/elasticsearch/gradle/DistributionDownloadPluginIT.java]: FWIW, a very common pattern in Gradle integration tests is to do these assertions in a task action in the build. That way your test just tests that a particular task ran successfully and you can avoid all this awkward business with passing info between the build and the test with raw output that needs to be parsed with regular expressions.

All this could be replaced with `assert file("${elasticsearch_distributions.test_distro}").exists()` in the task action.

File: buildSrc/src/testKit/jdk-download/settings.gradle [hunk 7]
```diff
@@ -1 +1,2 @@
-include 'subproj'
\ No newline at end of file
+include 'subproj'
+include 'reuse'
```
Comment (mark-vieira) [inline buildSrc/src/testKit/jdk-download/settings.gradle]: This seems unrelated.

**Review by mark-vieira — CHANGES_REQUESTED** [review]
Mostly minor comments, otherwise this looks great. Although it somewhat out of scope here, it would be nice to see what a "converted" project would look like, mostly to get an idea for real usage and DSL purposes. Is there a place we could do this in isolation or is that a larger, more involved piece of work. Mainly I just like to see that "minus" line count go up 😉

File: buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+    public void apply(Project project) {
+        distributionsContainer = project.container(ElasticsearchDistribution.class, name -> new ElasticsearchDistribution(name, project));
+        project.getExtensions().add("elasticsearch_distributions", distributionsContainer);
+
+        setupDownloadServiceRepo(project);
+
+        ExtraPropertiesExtension extraProperties = project.getExtensions().getExtraProperties();
+        if (extraProperties.has("bwcVersions")) {
+            this.bwcVersions = (BwcVersions) extraProperties.get("bwcVersions");
+        } // else - leniency for external plugins...TODO: setup snapshot dependency instead
```
[the remark this answers is not in the document]
Comment (rjernst) [reply buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java]: It's more than just bwcVersions. Actually, to do external projects correctly, we need bwcVersions to be present, so that we can know when a version is a snapshot. We would just go to the elastic snapshots service instead of using a local bwc project.

[the remark this answers is not in the document]
Comment (mark-vieira) [reply buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java]: I guess I don't quite follow. Should we error here then if that property doesn't exist?

File: buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+            distribution.finalizeValues();
+
+            DependencyHandler dependencies = project.getDependencies();
+            // for the distribution as a file, just depend on the artifact directly
+            dependencies.add(distribution.configuration.getName(), dependencyNotation(project, distribution));
+
+            // no extraction allowed for rpm or deb
+            if (distribution.getType().equals("rpm") == false && distribution.getType().equals("deb") == false) {
+                // for the distribution extracted, add a root level task that does the extraction, and depend on that
+                // extracted configuration as a fake artifact
```
[the remark this answers is not in the document]
Comment (rjernst) [reply buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java]: In this case, the coordinates are fake. I can see the confusion with calling it a fake artifact. But I do think eg FAKE_GROUP that is used here is correct; the group does not exist, it is artificial. I am happy to use another term, but I think it should convey the fact it is not an actual maven group, and that these coordinates cannot be looked up in a maven repository.

File: buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+        final ConfigurationContainer configurations = rootProject.getConfigurations();
+        String downloadConfigName = configName("elasticsearch", distribution);
+        String extractedConfigName = "extracted_" + downloadConfigName;
+        Configuration downloadConfig = configurations.findByName(downloadConfigName);
+        if (downloadConfig == null) {
+            downloadConfig = configurations.create(downloadConfigName);
+            configurations.create(extractedConfigName);
+        }
+        Object distroDep = dependencyNotation(rootProject, distribution);
+        rootProject.getDependencies().add(downloadConfigName, distroDep);
```
[the remark this answers is not in the document]
Comment (rjernst) [reply buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java]: Yes, that makes sense, since the configurations map one to one for a unique distribution.

File: buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+            throw new GradleException("no project [" + projectPath + "], project names: " + project.getRootProject().getAllprojects());
+        }
+        Map<String, Object> depConfig = new HashMap<>();
+        depConfig.put("path", projectPath);
+        depConfig.put("configuration", projectConfig);
+        return project.getDependencies().project(depConfig);
+    }
+
+    private static String distributionProjectPath(ElasticsearchDistribution distribution) {
+        String projectPath = ":distribution";
```
[the remark this answers is not in the document]
Comment (rjernst) [reply buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java]: Right now using from an external project will fail miserably. But I'll add this protection here as you suggested.

[the remark this answers is not in the document]
Comment (rjernst) [reply buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java]: However, I'm putting it in `projectDependency`, since that is what actually adds the dep.

[the remark this answers is not in the document]
Comment (rjernst) [reply buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java]: Actually, we sort of already have this protection, indirectly. I have this currently:

```
if (project.findProject(projectPath) == null) {
            throw new GradleException("no project [" + projectPath + "], project names: " + project.getRootProject().getAllprojects());
        }
```

Its not the exact same as what you suggested, but it does get the same result, as long as the external project is not matching our project structure. I can improve this further when adding support in a followup for external projects.

File: buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+        }
+        return projectName;
+    }
+
+    private static String configName(String prefix, ElasticsearchDistribution distribution) {
+        return prefix + "_" + distribution.getVersion() + "_" + distribution.getPlatform() + "_" + distribution.getType() + "_"
+            + distribution.getFlavor() + (distribution.getBundledJdk() ? "" : "_nojdk");
+    }
+
+    private static String capitalize(String s) {
```
[the remark this answers is not in the document]
Comment (rjernst) [reply buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java]: I'd like to leave that as a future cleanup. I'm not against it, but would like to minimize expanding the scope of this PR more, as I would want such a utils class to have tests.

File: buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java (hunk 1 above)
Comment (rjernst) [reply buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java]: Yes it is technically ok. I'd rather keep it strict here, where for rpm/deb a platform is not specified.

File: buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java (hunk 2 above)
Comment (rjernst) [reply buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java]: I thought about this at first, but that has a ton more methods to delegate to, and all we need is the iterable for now.

File: buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (mark-vieira) [reply buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java]: That's fine, but I'd really like us to become more open to these kinds of minor refactorings in the course of PRs. There are countless "we can do this in a separate PR" comments that are never followed up on because they are small and low priority. In practice, I fear this stuff just never gets done. I'd prefer untested DRY code to untested duplicate code.

File: buildSrc/src/test/java/org/elasticsearch/gradle/DistributionDownloadPluginIT.java (hunk 4 above)
Comment (rjernst) [reply buildSrc/src/test/java/org/elasticsearch/gradle/DistributionDownloadPluginIT.java]: These tests are quite slow, relatively, so i want to reduce the number of testkit builds we run. Currently the way this is structured, we run one within each assert method. I have a TODO at the top of this file to change that so the assertion for expanded checks both, because if we can use expanded, the file version should also exist.

Comment (mark-vieira) [reply buildSrc/src/test/java/org/elasticsearch/gradle/DistributionDownloadPluginIT.java]: Right, this test contains two assertions, so two testkit builds. Splitting this into two tests won't incur any additional build executions so I think it makes sense to split it up, unless we combine both assertions into a single build execution.

File: buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (rjernst) [reply buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java]: I removed the nullability. This was at first necessary for the integ tests which weren't always adding the bwcVersions, but is fixed now.

**Comment by rjernst** [comment]
@mark-vieira I believe I have addressed all your feedback.

**Review by mark-vieira — APPROVED** [review]

**Comment by rjernst** [comment]
@elasticmachine run elasticsearch-ci/1

File: buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+            // fall through: register the task
+        }
+        setupDownloadServiceRepo(rootProject);
+
+        final ConfigurationContainer configurations = rootProject.getConfigurations();
+        String downloadConfigName = configName("elasticsearch", distribution);
+        String extractedConfigName = "extracted_" + downloadConfigName;
+        final Configuration downloadConfig = configurations.create(downloadConfigName);
+        configurations.create(extractedConfigName);
+        Object distroDep = dependencyNotation(rootProject, distribution);
```
Comment (atorok) [inline buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java]: nit: lazy variable, used only once.

File: buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+        rootProject.getDependencies().add(downloadConfigName, distroDep);
+
+        // add task for extraction, delaying resolving config until runtime
+        if (distribution.getType() == Type.ARCHIVE || distribution.getType() == Type.INTEG_TEST_ZIP) {
+            Supplier<File> archiveGetter = downloadConfig::getSingleFile;
+            String extractDir = rootProject.getBuildDir().toPath().resolve("elasticsearch-distros").resolve(extractedConfigName).toString();
+            TaskProvider<Copy> extractTask = rootProject.getTasks().register(extractTaskName, Copy.class, copyTask -> {
+                copyTask.dependsOn(downloadConfig);
+                copyTask.doFirst(t -> rootProject.delete(extractDir));
```
Comment (atorok) [inline buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java]: Looks like we could use a  `Sync` task instead

File: buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+                copyTask.into(extractDir);
+                copyTask.from((Callable<FileTree>)() -> {
+                    File archiveFile = archiveGetter.get();
+                    String archivePath = archiveFile.toString();
+                    if (archivePath.endsWith(".zip")) {
+                        return rootProject.zipTree(archiveFile);
```
Comment (atorok) [inline buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java]: maybe it's just me, but I think `rootProject.zipTree(downloadConfig.getSingleFile())` would be much easier to read.  this might be subjective, but I find that variables only used once or a few times but really far apart make the code significantly harder to read.

File: buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java [hunk 16]
```diff
@@ -0,0 +1,273 @@
...
+            return "org.elasticsearch.distribution.integ-test-zip:elasticsearch:" + distribution.getVersion();
+        }
+
+        String extension = distribution.getType().toString();
+        String classifier = "x86_64";
+        if (distribution.getType() == Type.ARCHIVE) {
+            extension = distribution.getPlatform() == Platform.WINDOWS ? "zip" : "tar.gz";
+            classifier = distribution.getPlatform() + "-" + classifier;
+        }
+        return FAKE_GROUP + ":elasticsearch" + (distribution.getFlavor() == Flavor.OSS ? "-oss:" : ":")
```
Comment (atorok) [inline buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java]: `s/FAKE_GROUP.FAKE_IVY_GROUP/` might make it more obvious where this is coming from

File: buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+
+    private final Property<Version> version;
+    private final Property<Type> type;
+    private final Property<Platform> platform;
+    private final Property<Flavor> flavor;
+    private final Property<Boolean> bundledJdk;
+
+    ElasticsearchDistribution(String name, Project project) {
+        this.name = name;
+        this.configuration = project.getConfigurations().create("es_distro_file_" + name);
```
Comment (atorok) [inline buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java]: Can we avoid mutating the Gradle model in constructors ?

**Review by alpar-t — COMMENTED** [review]
I love this!  So glad to have it consolidated in a plugin.

File: buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java (hunk 14 above)
Comment (mark-vieira) [reply buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java]: We've run into issue with this before because the distributions include a bundled JDK which has odd file permissions.

Comment (atorok) [reply buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java]: Maybe add a comment explaining that, so we don't loose track of it ?

File: buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java [hunk 18]
```diff
@@ -0,0 +1,273 @@
...
+
+    @Override
+    public void apply(Project project) {
+        distributionsContainer = project.container(ElasticsearchDistribution.class, name -> new ElasticsearchDistribution(name, project));
+        project.getExtensions().add("elasticsearch_distributions", distributionsContainer);
+
+        setupDownloadServiceRepo(project);
+
+        ExtraPropertiesExtension extraProperties = project.getExtensions().getExtraProperties();
+        this.bwcVersions = (BwcVersions) extraProperties.get("bwcVersions");
```
[the remark this answers is not in the document]
Comment (rjernst) [reply buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java]: Baking it in is the plan. This was a first step.

File: buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java [hunk 19]
```diff
@@ -0,0 +1,229 @@
...
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.provider.Property;
+import org.gradle.api.tasks.TaskDependency;
+
+import java.io.File;
+import java.util.Iterator;
+import java.util.Locale;
+
+public class ElasticsearchDistribution implements Buildable {
```
[the remark this answers is not in the document]
Comment (rjernst) [reply buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java]: I think that class will go away once I convert test clusters to use this distribution plugin.

File: buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java [hunk 20]
```diff
@@ -0,0 +1,229 @@
...
+        DARWIN;
+
+        @Override
+        public String toString() {
+            return super.toString().toLowerCase(Locale.ROOT);
+        }
+    }
+
+    public enum Type {
+        INTEG_TEST_ZIP,
```
[the remark this answers is not in the document]
Comment (rjernst) [reply buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java]: I don't like that because then specifying it in build.gradle files requires doing an import. As it is now, all these values can be set by string.

File: buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java (hunk 15 above)
Comment (rjernst) [reply buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java]: I had something like that before, but Mark suggested switching on the extension. I don't have a strong preference, other than this is already in and works.

File: buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java [hunk 21]
```diff
@@ -0,0 +1,229 @@
...
+import org.gradle.api.provider.Property;
+import org.gradle.api.tasks.TaskDependency;
+
+import java.io.File;
+import java.util.Iterator;
+import java.util.Locale;
+
+public class ElasticsearchDistribution implements Buildable {
+
+    public enum Platform {
```
[the remark this answers is not in the document]
Comment (rjernst) [reply buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java]: We have a couple Platform like classes. I'd like to consider consolidation separately.

**Comment by rjernst** [comment]
@atorok I opened https://github.com/elastic/elasticsearch/pull/43461 to address some of your comments.

File: buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java (hunk 15 above)
Comment (mark-vieira) [reply buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java]: We were switching based on hard-coded distribution types. That certainly worked but is more brittle than just looking at the actual file-extension. That also was using raw strings before the conversion to enums which was even more brittle. Other option is to have some property on the distribution to indicate what _kind_ of archive it is and key off of that.

File: buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java (hunk 17 above)
Comment (mark-vieira) [reply buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java]: Agreed. This is not best practice. We should probably create the configuration in the plugin and pass as a constructor argument. We could also then avoid passing `Project` altogether and instead inject an `ObjectFactory` for all the property creation.

File: buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (atorok) [reply buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java]: @rjernst 
Not if we make that the default :) Or use some Groovy magic to avoid the import ? 
It's not a big deal, but calling it a type feels wrong  and I think it forces us to do more conditionals down the road.

[the remark this answers is not in the document]
Comment (mark-vieira) [reply buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java]: One option is to provide a convenience method on the `ElasticsearchDistribution` container similar to somthing like `mavenCentral()` on `RepositoryHandler`.

```
elasticsearch_distributions {
    integTest()
}```

It just now also occurred to me that we use snake case for this extension name. Thoughts on making this camel case as is convention with most Gradle plugin extensions?
